src/rust/engine/src/scheduler.rs

```diff
@@ -384,6 +385,9 @@ impl Scheduler {
           }
           res = &mut execution_task => {
             // Completed successfully.
+            // Wait for tail tasks to complete.
+            let tail_tasks = session.tail_tasks().lock().drain(..).collect();
+            Self::wait_for_tail_tasks(tail_tasks).await;
```

Comment:
Most likely we'd want to do this at shutdown of the `StreamingWorkunitHandler` instead, which is also where #11433 comes into play: https://github.com/pantsbuild/pants/blob/e4c1fc37725968147c2e559d5f200fffc66931f9/src/python/pants/bin/local_pants_runner.py#L260-L271 ... essentially, the `StreamingWorkunitHandler` would be finalized in a the background, and would wait for tail tasks before doing so.

Reply:
I'm fine with doing the wait wherever it makes better sense to do so. This PR just contains my pure guess as to where to do that.